`kube/proxy.TestModeratedSessions` race

## Failure

#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/4056205714/jobs/6980390339

#### Relevant snippet

```
===================================================
OUTPUT github.com/gravitational/teleport/lib/kube/proxy.TestModeratedSessions
===================================================
=== RUN   TestModeratedSessions
time="2023-01-31T16:36:44Z" level=warning msg="Mux serve failed." error="listener is closed\n\taccept tcp 127.0.0.1:34343: use of closed network connection" trace.component=auth
==================
WARNING: DATA RACE
Write at 0x00c001e2e7d0 by goroutine 1913:
  github.com/gravitational/teleport/lib/kube/proxy.(*TLSServer).Serve()
      /__w/teleport/teleport/lib/kube/proxy/server.go:251 +0x4a4
  github.com/gravitational/teleport/lib/kube/proxy.(*testContext).startKubeService.func1()
      /__w/teleport/teleport/lib/kube/proxy/utils_test.go:244 +0x70

Previous read at 0x00c001e2e7d0 by goroutine 1849:
  github.com/gravitational/teleport/lib/kube/proxy.(*TLSServer).Close()
      /__w/teleport/teleport/lib/kube/proxy/server.go:269 +0x286
  github.com/gravitational/teleport/lib/kube/proxy.(*testContext).Close()
      /__w/teleport/teleport/lib/kube/proxy/utils_test.go:256 +0x55
  github.com/gravitational/teleport/lib/kube/proxy.TestHeartbeat.func2()
      /__w/teleport/teleport/lib/kube/proxy/server_test.go:260 +0x39
  testing.(*common).Cleanup.func1()
      /opt/go/src/testing/testing.go:1041 +0x193
  testing.(*common).runCleanup()
      /opt/go/src/testing/testing.go:1210 +0x143
  testing.tRunner.func2()
      /opt/go/src/testing/testing.go:1440 +0x52
  runtime.deferreturn()
      /opt/go/src/runtime/panic.go:476 +0x32
  testing.(*T).Run.func1()
      /opt/go/src/testing/testing.go:1493 +0x47

Goroutine 1913 (running) created at:
  github.com/gravitational/teleport/lib/kube/proxy.(*testContext).startKubeService()
      /__w/teleport/teleport/lib/kube/proxy/utils_test.go:243 +0x197
  github.com/gravitational/teleport/lib/kube/proxy.setupTestContext()
      /__w/teleport/teleport/lib/kube/proxy/utils_test.go:229 +0x1a57
  github.com/gravitational/teleport/lib/kube/proxy.TestHeartbeat()
      /__w/teleport/teleport/lib/kube/proxy/server_test.go:252 +0x346
  testing.tRunner()
      /opt/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /opt/go/src/testing/testing.go:1493 +0x47

Goroutine 1849 (finished) created at:
  testing.(*T).Run()
      /opt/go/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /opt/go/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /opt/go/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /opt/go/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /opt/go/src/testing/testing.go:1726 +0xa84
  main.main()
      _testmain.go:177 +0x3bc
==================
time="2023-01-31T16:36:44Z" level=info msg="Generating TLS certificate {0xc00045df20 0xc000e96080 1.3.9999.1.7=#1310726f6f742e6578616d706c652e[636](https://github.com/gravitational/teleport/actions/runs/4056205714/jobs/6980390339#step:6:637)f6d,CN=test-tls-server.root.example.com,O=Auth,POSTALCODE=null,STREET= 2033-01-28 16:36:44.285090911 +0000 UTC [*.teleport.cluster.local teleport.cluster.local] [] 5 []}." common_name=test-tls-server.root.example.com dns_names="[*.teleport.cluster.local teleport.cluster.local]" locality="[]" not_after="2033-01-28 16:36:44.285090911 +0000 UTC" org="[Auth]" org_unit="[]" trace.component=ca
time="2023-01-31T16:36:44Z" level=info msg="Generating TLS certificate {0xc00045df20 0xc000e6e480 1.3.9999.1.7=#1310726f6f742e6578616d706c[652](https://github.com/gravitational/teleport/actions/runs/4056205714/jobs/6980390339#step:6:653)e636f6d,CN=0b3c3d7a-9809-48ac-94ee-3057ad545859.root.example.com,O=Kube,POSTALCODE=null,STREET= 2033-01-28 16:36:44.285090911 +0000 UTC [*.teleport.cluster.local teleport.cluster.local] [] 5 []}." common_name=0b3c3d7a-9809-48ac-94ee-3057ad545859.root.example.com dns_names="[*.teleport.cluster.local teleport.cluster.local]" locality="[]" not_after="2033-01-28 16:36:44.285090911 +0000 UTC" org="[Kube]" org_unit="[]" trace.component=ca
time="2023-01-31T16:36:45Z" level=info msg="Generating TLS certificate {0xc00045df20 0xc000f40030 1.3.9999.1.7=#1310726f6f742e[657](https://github.com/gravitational/teleport/actions/runs/4056205714/jobs/6980390339#step:6:658)8616d706c652e636f6d,CN=Proxy.root.example.com,O=Proxy,POSTALCODE=null,STREET= 2033-01-28 16:36:44.285090911 +0000 UTC [*.teleport.cluster.local teleport.cluster.local] [] 5 []}." common_name=Proxy.root.example.com dns_names="[*.teleport.cluster.local teleport.cluster.local]" locality="[]" not_after="2033-01-28 16:36:44.285090911 +0000 UTC" org="[Proxy]" org_unit="[]" trace.component=ca
time="2023-01-31T16:36:45Z" level=info msg="Generating TLS certificate {0xc00045df20 0xc0016b88f0 1.3.9999.1.7=#1310726f6f742e6578616d706c652e636f6d,CN=0b3c3d7a-9809-48ac-94ee-3057ad545859.root.example.com,O=Auth,POSTALCODE=null,STREET= 2033-01-28 16:36:44.285090911 +0000 UTC [*.teleport.cluster.local teleport.cluster.local] [] 5 []}." common_name=0b3c3d7a-9809-48ac-94ee-3057ad545859.root.example.com dns_names="[*.teleport.cluster.local teleport.cluster.local]" locality="[]" not_after="2033-01-28 16:36:44.285090911 +0000 UTC" org="[Auth]" org_unit="[]" trace.component=ca
=== CONT  TestModeratedSessions
    testing.go:1319: race detected during execution of test
--- FAIL: TestModeratedSessions (1.58s)
===================================================
```

--- a/lib/kube/proxy/server.go
+++ b/lib/kube/proxy/server.go
@@ -269,7 +269,9 @@ func (t *TLSServer) Close() error {
 	if t.watcher != nil {
 		t.watcher.Close()
 	}
+	t.mu.Lock()
 	listClose := t.listener.Close()
+	t.mu.Unlock()
 	return trace.NewAggregate(append(errs, listClose)...)
 }
 

